### ISSUE TITLE
system_traces.events gets populated even though tracing has been disabled

### ISSUE THREAD

**Issue by gibsn** [scylladb/scylla#6014]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [*] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

Installation details
Scylla version (or git commit hash): 3.2.2
Cluster size: 9 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): C7.5

Platform (physical/VM/cloud instance type/docker): bare metal
Hardware: sockets=2 x Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz cores=40 hyperthreading=yes memory= 6x32GB DDR4 2666MHz
Disks: RAID 10 of 10HDDs 14TB each for data, RAID 1 SSD 1TB for clogs

 If during the same session you have tracing enabled first, then disable, you will still get the `events` table populated with entries corresponding to the second query with the same `session_id`. In our case we started experiencing huge timeouts after updating to 3.2.2 (#5922), probably  because tracings table gets overloaded.

The issue has been spotted using the `gocql` lib, but then was reproduced using `cqlsh`. Steps to reproduce:
```
truncate system_traces.events; truncate system_traces.sessions;
tracing on;
// run some query
// check count(*) events
tracing off;
// run some query
// check count(*) events, see growth
```

Please also check the contents of `events` and `sessions` tables:
[events.txt](https://github.com/scylladb/scylla/files/4326922/events.txt)
[sessions.txt](https://github.com/scylladb/scylla/files/4326924/sessions.txt)

You can see that `event_id` in `events` differs in seconds, while there is only one unique `session_id`

**Comment by slivne** [comment]
probably related to scylladb/scylla#5700 and (https://github.com/scylladb/scylla/issues/5700#issuecomment-586706321),

https://github.com/scylladb/scylla/commit/866c04dd6440cb51e752dd072019426aa5192659

**Comment by slivne** [comment]
I checked on 3.3.rc3 the issue indeed seem to be resolved using the simple reproducer provided above

**Comment by slivne** [comment]
@eliransin you don;t need todo anything here - we need todo a backport when we think the code in 866c04d is stable enough probably after 3.3 goes out

**Comment by slivne** [comment]
I am removing assignment

**Comment by slivne** [comment]
@avikivity / @gleb-cloudius - this is the issue related to tracing

**Comment by gleb-cloudius** [comment]
I would be surprised if #5700 is indeed to blame. When tracing is off no tracing ptr should be created at all, so what I would expect with the bug disabling tracing would cause some entries that should be there no be, but not other way around. 

OTOH I am not sure the issue as described is an issue at all. If some queries that were run before turning off the tracing are still running in the background they will add entries to event table. If this is the case there is nothing to fix here.

**Comment by slivne** [comment]
@Gleb Natapov <gleb@scylladb.com> I ran the reproducer the user provided -
its not background queries its new queries that are logged.

**Comment by gleb-cloudius** [comment]
Yes, I see it in the provided file. Looks wrong indeed.

**Comment by vladzcloudius** [comment]
@gleb-cloudius @slivne The regression has been introduced by 
```
commit f78b2c558895f8f9704639c4f14612ba791a3b7e
Author: Gleb Natapov <gleb@scylladb.com>
Date:   Fri Sep 6 11:30:50 2019 +0300

    transport: remove remaining craft related to cql's server load balancing
    
    Commit 7e3805ed3ded327ff removed the load balancing code from cql
    server, but it did not remove most of the craft that load balancing
    introduced. The most of the complexity (and probably the main reason the
    code never worked properly) is around service::client_state class which
    is copied before been passed to the request processor (because in the past
    the processing could have happened on another shard) and then merged back
    into the "master copy" because a request processing may have changed it.
    
    This commit remove all this copying. The client_request is passed as a
    reference all the way to the lowest layer that needs it and it copy
    construction is removed to make sure nobody copies it by mistake.
    
    tests: dev, default c-s load of 3 node cluster
    
    Message-Id: <20190906083050.GA21796@scylladb.com>

```
Until that patch a new instance of `client_state` was created for each `cql_server::connection::process_request_one(...)` call and passing of tracing state in it was totally safe.

The patch above broke that invariant.

**Comment by gleb-cloudius** [comment]
That's good news since this is what 866c04d fixes, though I still not see how the bug can prolong trace_ptr live, only shorten it.

**Comment by vladzcloudius** [comment]
> That's good news since this is what [866c04d](https://github.com/scylladb/scylla/commit/866c04dd6440cb51e752dd072019426aa5192659) fixes, though I still not see how the bug can prolong trace_ptr live, only shorten it.

@gleb-cloudius You mean, "how the patch can prolong", right?
That's exactly as in the description of 866c04d: trace_ptr is stored in the client_state and since the lifetime of the client_state is no longer bound to the life time of the current request handling it remains alive (since it's referenced from the shared client_state, which is not a field in the cql_server::connection).

As a result the trace_ptr created for the previously traced query is returned for all following not traced (!!!) queries.

If the following query IS traced too then the old trace_ptr is being overwritten and finally destroyed but before it does it keeps on generating new tracing events generating a huge partition in `system_traces.events` table.

After it is replaced by the new trace_ptr the new trace_ptr starts generating a huge partition and so on.

Namely once you traced a single query - you are doomed.

**Comment by gleb-cloudius** [comment]
Right. For some reason I thought that when you disable tracing a pointer
in client_state is overwritten with nullptr, but looking at the code it does
not.

--
			Gleb.

**Comment by slivne** [comment]
@gleb-cloudius since you are familiar with the code whats the easy way to fix this in 3.2 - is there a simple backport ?

**Comment by gleb-cloudius** [comment]
I do not understand what is the difference between this an any other
issue that is already fixed in master. The easiest way is to do a
backport of the fix like we always do. If maintainers have issues they
can always ask for help.

--
			Gleb.

**Comment by eliransin** [comment]
@vladzcloudius  FYI :+1:  https://github.com/scylladb/scylla/issues/6014#event-3157262031

**Comment by slivne** [comment]
fixed by a092f5d1f4f1b39c3e3734df00bd30e91b6d4282

will be part pf 3.2.4
